crates/kitsune_p2p/kitsune_p2p/src/spawn/actor/space/bootstrap_task.rs

```diff
@@ -0,0 +1,708 @@
+    max_delay: Duration,
+}
+
+// Trait for the bootstrap query to allow mocking in tests
+trait BootstrapService: Send {
+    fn random(&self, query: RandomQuery) -> BoxFuture<KitsuneP2pResult<Vec<AgentInfoSigned>>>;
+}
+
+struct DefaultBootstrapService {
+    url: Option<Url2>,
+    net: BootstrapNet,
+}
+
+impl BootstrapService for DefaultBootstrapService {
+    fn random(&self, query: RandomQuery) -> BoxFuture<KitsuneP2pResult<Vec<AgentInfoSigned>>> {
+        super::bootstrap::random(self.url.clone(), query, self.net).boxed()
+    }
+}
+
+impl BootstrapTask {
+    pub(super) fn spawn(
+        internal_sender: GhostSender<SpaceInternal>,
+        host_sender: Sender<KitsuneP2pEvent>,
+        space: Arc<KitsuneSpace>,
+        bootstrap_service: Option<Url2>,
+        bootstrap_net: BootstrapNet,
+        bootstrap_check_delay_backoff_multiplier: u32,
+    ) -> Arc<RwLock<Self>> {
+        let this = Arc::new(RwLock::new(BootstrapTask {
+            is_finished: false,
+            current_delay: Duration::from_secs(1),
+            max_delay: Duration::from_secs(60 * 60),
+        }));
+        let bootstrap_query = DefaultBootstrapService {
+            url: bootstrap_service,
+            net: bootstrap_net,
+        };
+
+        let task_this = this.clone();
+        BootstrapTask::spawn_inner(
+            task_this,
+            internal_sender,
+            host_sender,
+            space,
+            Box::new(bootstrap_query),
+            bootstrap_check_delay_backoff_multiplier,
+        )
+    }
+
+    fn spawn_inner(
+        this: Arc<RwLock<Self>>,
+        internal_sender: GhostSender<SpaceInternal>,
+        host_sender: Sender<KitsuneP2pEvent>,
+        space: Arc<KitsuneSpace>,
+        bootstrap_query: Box<impl BootstrapService + Send + Sync + 'static>,
+        bootstrap_check_delay_backoff_multiplier: u32,
+    ) -> Arc<RwLock<Self>> {
+        let task_this = this.clone();
+        tokio::spawn(async move {
+            let backoff_multiplier = if bootstrap_check_delay_backoff_multiplier < 2 {
+                tracing::warn!(
+                    "Using default bootstrap backoff multiplier 2 because configured value is too low - {}",
+                    bootstrap_check_delay_backoff_multiplier
+                );
+                2
+            } else {
+                bootstrap_check_delay_backoff_multiplier
+            };
+
+            let max_delay = task_this.read().max_delay;
+
+            loop {
+                if !internal_sender.ghost_actor_is_active() {
+                    break;
+                }
+
+                let current_delay = task_this.read().current_delay;
+                tokio::time::sleep(current_delay).await;
+                if current_delay <= max_delay {
+                    // Backoff but don't exceed the configured max delay
+                    task_this.write().current_delay =
+                        std::cmp::min(current_delay * backoff_multiplier, max_delay);
```

Comment:
This is changed, there was no check previously that the value stayed within the max. Even with the default values of a 1s delay and 2x delay multiplier you would actually end up with a max of 64 minutes, not 60 as intended. The max is now respected even with high values for the delay multiplier